src/main/scala/com/fulcrumgenomics/bam/AnnotateMolecularBarcodes.scala

```diff
@@ -0,0 +1,148 @@
+  @arg(flag = "a",   doc = "Annotate the read names with the molecular barcodes.") val annotateReadNames: Boolean = true,
+  @arg(flag = "t",   doc = "Annotate both reads from a read pair with each others tags.")
+                                                                                  val tagBothEnds: Boolean = true
+) extends FgBioTool with LazyLogging {
+
+  Io.assertReadable(input)
+  Io.assertCanWriteFile(output)
+  molecularBarcodeTags1.foreach(tag => if (tag.length != 2) throw new ValidationException("SAM tags must be of length two: " + tag))
+  molecularBarcodeTags2.foreach(tag => if (tag.length != 2) throw new ValidationException("SAM tags must be of length two: " + tag))
+  if (readStructure1.isEmpty) throw new ValidationException("readStructure1 may not be empty")
+  if (readStructure2.isEmpty) throw new ValidationException("readStructure2 may not be empty")
+  if (tagBothEnds) {
+    molecularBarcodeTags1.foreach { tag => assert(!molecularBarcodeTags2.contains(tag)) }
+    molecularBarcodeTags2.foreach { tag => assert(!molecularBarcodeTags1.contains(tag)) }
+  }
+
+  override def execute(): Unit = {
+    val rs1 = ReadStructure(readStructure1)
+    val rs2 = ReadStructure(readStructure2)
+    assert(molecularBarcodeTags1.length == rs1.count {
+      case m: MolecularBarcode => true
+      case _ => false
+    })
+    assert(molecularBarcodeTags2.length == rs2.count {
+      case m: MolecularBarcode => true
+      case _ => false
+    })
+    val in: SamReader = SamReaderFactory.make.open(input.toFile)
+    val iterator: SAMRecordIterator = in.iterator()
+    val out: SAMFileWriter = new SAMFileWriterFactory().makeSAMOrBAMWriter(in.getFileHeader, true, output.toFile)
+    iterator.sliding(2,2).foreach {
+      case Seq(r1: SAMRecord, r2: SAMRecord) =>
+        assert(r1.getReadName.compareTo(r2.getReadName) == 0)
+        val bases1 = annotateRecord(record=r1, readStructure=rs1, molecularBarcodeTags=molecularBarcodeTags1)
+        val bases2 = annotateRecord(record=r2, readStructure=rs2, molecularBarcodeTags=molecularBarcodeTags2)
+        if (annotateReadNames) {
+          r1.setReadName(r1.getReadName + "|" + bases1 + bases2)
+          r2.setReadName(r2.getReadName + "|" + bases1 + bases2)
+        }
+        if (tagBothEnds) {
+          molecularBarcodeTags1.foreach { tag => r2.setAttribute(tag, r1.getStringAttribute(tag)) }
+          molecularBarcodeTags2.foreach { tag => r1.setAttribute(tag, r2.getStringAttribute(tag)) }
+        }
+        out.addAlignment(r1)
+        out.addAlignment(r2)
+    }
+    out.close()
+    CloserUtil.close(iterator)
+  }
+
+  /**
+    * Extracts bases associated with molecular barcodes and adds them as SAM tags, and optionally concatenates
+    * them to the read name.  The read's bases will only contain template bases as defined in the given read structure.
+    */
+  private def annotateRecord(record: SAMRecord,
```

Comment:
Agreed someone has to modify - it's the modify _and_ return something else that I'm objecting to.  I might be wrong on instance state since I can no longer see what I was talking about either ;)